app/web/features/auth/notifications/PushNotificationPermission.tsx

```diff
@@ -0,0 +1,193 @@
+      setIsPushEnabled(
+        Notification.permission === "granted" &&
+          existingPushSubscription !== null
+      );
+    };
+
+    checkPushEnabled();
+    setIsLoading(false);
+  }, []);
+
+  const onPermissionGranted = async () => {
+    try {
+      // Check if service workers and push notifications are supported
+      if ("serviceWorker" in navigator && "PushManager" in window) {
+        const existingPushSubscription = await getCurrentSubscription();
+        const p256dhKey = existingPushSubscription?.getKey("p256dh");
+        const { vapidPublicKey } = await getVapidPublicKey();
+
+        if (existingPushSubscription && p256dhKey) {
+          const publicKey = arrayBufferToBase64(p256dhKey);
+
+          /**
+           * The purpose of this check is to ensure that the push subscription is correctly authenticated with the server’s VAPID key.
+           * If the client’s p256dh key no longer matches the server’s vapidPublicKey, then the subscription is unsubscribed and needs
+           * to be re-registered to ensure the security and validity of the Web Push connection.
+           */
+          if (publicKey !== vapidPublicKey) {
+            await existingPushSubscription.unsubscribe();
+          } else {
+            return;
+          }
+        }
+
+        const registration = await navigator.serviceWorker.getRegistration();
+
+        // Subscribe to push notifications via the PushManager
+        const subscription: PushSubscription =
+          await registration!.pushManager.subscribe({
+            userVisibleOnly: true,
+            applicationServerKey: vapidPublicKey,
+          });
+
+        await registerPushNotificationSubscription(subscription);
+      } else {
+        setErrorMessage(
+          t("notification_settings.push_notifications.error_unsupported")
+        );
+        Sentry.captureException(
+          new Error("Push notifications or service workers not supported"),
+          {
+            tags: {
+              component: "PushNotificationPermission",
+              action: "onPermissionGranted",
+              userAgent: navigator.userAgent,
+            },
+          }
+        );
+      }
+    } catch (error) {
+      console.error("Error subscribing to push notifications", error);
+      setErrorMessage(
+        t("notification_settings.push_notifications.error_generic")
+      );
+
+      Sentry.captureException(error, {
+        tags: {
+          component: "PushNotificationPermission",
+          action: "onPermissionGranted",
+        },
+      });
+    }
+  };
+
+  const turnPushNotificationsOn = async () => {
+    if (Notification.permission !== "denied") {
+      setIsLoading(true);
+      setShouldPromptAllow(true);
+      const result = await Notification.requestPermission();
+      setShouldPromptAllow(false);
+
+      if (result === "granted") {
+        await onPermissionGranted();
+        setIsPushEnabled(true);
+      }
```

Comment:
Should this have an else with `setIsPushEnabled(false)`?

Reply:
Yeah makes sense.